contrib_bots/lib/gmail.py

```diff
@@ -0,0 +1,163 @@
+import logging
+import mimetypes
+import os
+import re
+
+from googleapiclient import discovery, errors
+from email.mime.text import MIMEText
+from oauth2client.file import Storage
+
+class GmailHandler(object):
+    '''
+    This plugin facilitates sending emails from a Gmail account
+    from within Zulip. It looks for messages starting with '@gmail'.
+    '''
+
+    def usage(self):
+        return '''
+            This plugin will allow users to send emails from a
+            personal gmail account. Users should preface
+            messages with "@gmail".
+
+            Enter messages in the following format:
+            '@gmail <recipient_email> "<subject>" "<body>" --label <label>'
+            The --label tag is optional and can be left out.
+            '''
+
+    def triage_message(self, message, client):
+        # return True if we want to (possibly) response to this message
+
+        original_content = message['content']
+        self.bot_email = client.email
+
+        # This next line of code is defensive
+        if message['display_recipient'] == 'gmail':
+            return False
+        return original_content.startswith('@gmail')
+
+    def handle_message(self, message, client, state_handler):
+        original_content = message['content']
+        original_sender = message['sender_email']
+
+        try:
+            recipient, subject, body, label = parse_input(original_content)
+        except ValueError:
+            recipient, subject, body = parse_input(original_content)
+            label = False
+        except TypeError:
+            raise Exception("The input format used was incorrect. The correct format is: "
+                            '@gmail <recipient> "<subject>" "<body>"')
+
+        service = get_service()
+        email = create_message(body, subject, recipient)
+        result = send_message(service, email)
+
+        label_result = False
+        if label:
+            create_label(service, label)
+            label_id = get_label_id(service, label)
+            label_result = add_label(service, label_id, result['id'])
+
+        if 'SENT' in result['labelIds']:
+            content = 'Email sent to {}'.format(recipient)
+            if label_result:  # Enters this block if the label couldn't be added
+                              # or if there was no label
+                content += ' without label.'
+            else:
+                content += '.'
+        else:
+            content = 'Could not send email to {}. Please try again.'.format(recipient)
+
+        client.send_message(dict(
+            type = 'private',
+            to = original_sender,
+            sender = self.bot_email,
+            content = content
+        ))
+
+
+def parse_input(user_input):
+    # Input is of the following format: (--label flag is optional)
+    # @gmail <recipient_email> "<subject>" "<body>" --label <label>
+    if '--label' in user_input:
+        regex = '@gmail (.*?) "(.*?)" "(.*?)" --label (.*?)\Z'
```

Comment:
Is there any performance difference? If not I don't understand why it would matter.